File: src/sense_hat/imu_strategy_factory.py
from collections.abc import Callable
from dataclasses import dataclass

from builtin_interfaces.msg import Time

from sense_hat.imu_strategy_protocol import ImuStrategy
from sense_hat.stationary_imu_mock import StationaryImuMock
from sense_hat.udp_reader_imu_strategy import DatagramSocket, UdpReaderImuStrategy
from sense_hat.yaw_rotation_imu_mock import YawRotationImuMock

STRATEGY_MODE_UDP_READER = "udp_reader"
MOCK_MODE_STATIONARY = "stationary"
MOCK_MODE_YAW_ROTATION = "yaw_rotation"
MOCK_MODES = {MOCK_MODE_STATIONARY, MOCK_MODE_YAW_ROTATION}
STRATEGY_MODES = {STRATEGY_MODE_UDP_READER, *MOCK_MODES}


@dataclass(frozen=True, kw_only=True)
class ImuStrategyConfig:
    """Configuration for creating an IMU strategy."""

    mode: str
    frame_id: str
    stamp_factory: Callable[[], Time]
    host: str = "127.0.0.1"
    port: int = 8765
    angular_velocity_z: float = 0.25


def create_imu_strategy(config: ImuStrategyConfig, sock: DatagramSocket | None = None) -> ImuStrategy:
    """Create an IMU strategy from configuration."""
    if config.mode == STRATEGY_MODE_UDP_READER:
        return UdpReaderImuStrategy(
            host=config.host,
            port=config.port,
            frame_id=config.frame_id,
            stamp_factory=config.stamp_factory,
            sock=sock,
        )

    if config.mode == MOCK_MODE_STATIONARY:
        return StationaryImuMock(
            frame_id=config.frame_id,
            stamp_factory=config.stamp_factory,
        )

    if config.mode == MOCK_MODE_YAW_ROTATION:
        return YawRotationImuMock(
            frame_id=config.frame_id,
            stamp_factory=config.stamp_factory,
            angular_velocity_z=config.angular_velocity_z,
        )

    msg = f"Unsupported IMU strategy mode '{config.mode}'. Use one of {sorted(STRATEGY_MODES)}."
    raise ValueError(msg)

File: src/sense_hat_bridge/src/imu_strategy_factory.cpp
#include "sense_hat_bridge/imu_strategy_factory.hpp"

#include <stdexcept>
#include <string>

#include "sense_hat_bridge/stationary_imu_mock.hpp"
#include "sense_hat_bridge/udp_reader_imu_strategy.hpp"
#include "sense_hat_bridge/yaw_oscillation_imu_mock.hpp"
#include "sense_hat_bridge/yaw_rotation_imu_mock.hpp"

namespace sense_hat_bridge {

std::unique_ptr<ImuStrategy> create_imu_strategy(const ImuStrategyConfig& config) {
  if (config.mode == kStrategyModeUdpReader) {
    return std::make_unique<UdpReaderImuStrategy>(config.host, config.port, config.frame_id,
                                                  config.stamp_factory);
  }

  if (config.mode == kMockModeStationary) {
    return std::make_unique<StationaryImuMock>(config.frame_id, config.stamp_factory);
  }

  if (config.mode == kMockModeYawRotation) {
    return std::make_unique<YawRotationImuMock>(
        config.frame_id, config.stamp_factory,
        YawRotationImuMock::YawRotationRampParams{
            .start_angular_velocity_z = config.old_snapshot.angular_velocity_z,
            .end_angular_velocity_z = config.angular_velocity_z,
            .duration_s = 1.0});
  }

  if (config.mode == kMockModeYawOscillation) {
    return std::make_unique<YawOscillationImuMock>(
        config.frame_id, config.stamp_factory,
        YawOscillationImuMock::Params{
            .start_angular_velocity_z = config.old_snapshot.angular_velocity_z,
            .amplitude_angular_velocity_z = config.angular_velocity_z,
            .ramp_duration_s = YawOscillationImuMock::kDefaultRampDuration,
            .frequency_hz = config.yaw_oscillation_frequency_hz});
  }

  throw std::invalid_argument{"Unsupported IMU strategy mode '" + config.mode +
                              "'. Use one of: stationary, yaw_rotation, yaw_oscillation, "
                              "udp_reader."};
}

}  // namespace sense_hat_bridge
